proposals/type-classes.md

```diff
@@ -0,0 +1,298 @@
+# Type classes
+
+* **Type**: Design proposal
+* **Author**: Raul Raja
+* **Status**: New
+* **Prototype**: -
+
+## Summary
+
+The goal of this proposal is to enable `type classes` and lightweight `Higher Kinded Types` in Kotlin to enable ad-hoc polymorphism and better extension syntax.
+Type classes is the most important feature that Kotlin lacks in order to support a broader range of FP idioms.
+Kotlin already has an excellent extension mechanism where this proposal fits nicely. As a side effect `Type classes as extensions` also allows for compile time
+dependency injection which will improve the current landscape where trivial applications rely on heavy frameworks based on runtime Dependency Injection.
+Furthermore introduction of type classes improves usages for `reified` generic functions with a more robust approach that does not require those to be `inline` or `reified`.
+
+## Motivation
+
+* Support Type class evidence compile time verification.
+* Support a broader range of Typed FP patterns.
+* Enable multiple extension functions groups for type declarations.
+* Enable compile time DI through the use of the Type class pattern.
+* Enable better compile reified generics without the need for explicit inlining.
+* Enable definition of polymorphic functions whose constrains can be verified at compile time in call sites.
+
+## Description
+
+We propose to use the existing `interface` semantics allowing for generic definition of type classes and their instances with the same style interfaces are defined
+
+```kotlin
+interface Monoid<A> {
+    fun A.combine(b: A): A
+    fun empty(): A
+}
+```
+
+The above declaration can serve as target for implementations for any arbitrary data type.
+In the implementation below we provide evidence that there is a `Monoid<Int>` instance that enables `combine` and `empty` on `Int`
+
+```kotlin
+package intext
+
+object IntMonoid : Monoid<Int> {
```

Comment:
@abreslav thanks for the link, I'll check it out. It was an original part of the discussion that only objects or classes flagged as `extension` should be suitable candidates to become type class instances available for resolution. I agreed that those should be specially flagged to avoid abuse and confusion with regular sub-typing. I'll include that in the proposal.